core/services/keystore/keys/terrakey/key.go

```diff
@@ -0,0 +1,79 @@
+package terrakey
+
+import (
+	cryptorand "crypto/rand"
+	"fmt"
+	"io"
+
+	cosmosed25519 "github.com/cosmos/cosmos-sdk/crypto/keys/ed25519"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/mr-tron/base58"
+)
+
+type Raw []byte
+
+func (raw Raw) Key() Key {
+	return Key{
+		PrivKey: cosmosed25519.GenPrivKeyFromSecret(raw),
+		secret:  raw,
+	}
+}
+
+func (raw Raw) String() string {
+	return "<Terra Raw Private Key>"
+}
+
+func (raw Raw) GoString() string {
+	return raw.String()
+}
+
+var _ fmt.GoStringer = &Key{}
+
+type Key struct {
+	*cosmosed25519.PrivKey
+	secret []byte
+}
+
+func New() Key {
+	return newFrom(cryptorand.Reader)
+}
+
+func MustNewInsecure(reader io.Reader) Key {
+	return newFrom(reader)
+}
+
+func newFrom(reader io.Reader) Key {
+	secret := make([]byte, 32)
+	_, err := io.ReadFull(reader, secret)
+	if err != nil {
+		panic(err)
+	}
+	return Key{
+		PrivKey: cosmosed25519.GenPrivKeyFromSecret(secret),
+		secret:  secret,
+	}
+}
+
+func (key Key) ID() string {
+	return key.PublicKeyStr()
+}
+
+func (key Key) PublicKey() (pubKey cryptotypes.PubKey) {
+	return key.PubKey()
+}
+
+func (key Key) PublicKeyStr() string {
+	return base58.Encode(key.PubKey().Bytes())
```

Comment:
the `base58` encoding is just a solana thing not a cosmos/terra thing

i think here's a starting point in the terra-go sdk:
```
// Generate Address from Public Key
addr := msg.AccAddress(privKey.PubKey().Address())
assert.Equal(t, addr.String(), "terra1cevwjzwft3pjuf5nc32d9kyrvh5y7fp9havw7k")
```
https://github.com/terra-money/terra.go

it looks like it points back to some cosmos sdks :)

Reply:
ahh right :+1: good catch

Reply:
:+1: